csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -126,9 +127,28 @@ public void AddEntriesFrom(ref ParseContext ctx, FieldCodec<T> codec)
                 if (length > 0)
                 {
                     int oldLimit = SegmentedBufferHelper.PushLimit(ref ctx.state, length);
-                    while (!SegmentedBufferHelper.IsReachedLimit(ref ctx.state))
+
+                    // If the content is fixed size then we can calculate the length
+                    // of the repeated field and pre-initialize the underlying collection.
+                    //
+                    // Check that the supplied length doesn't exceed the underlying buffer.
+                    // That prevents a malicious length from initializing a very large collection.
+                    if (codec.FixedSize > 0 && IsSizeAvailable(ref ctx, length))
+                    {
+                        EnsureSize(count + (length / codec.FixedSize));
+
+                        while (!SegmentedBufferHelper.IsReachedLimit(ref ctx.state))
+                        {
+                            array[count++] = reader(ref ctx);

```

Comment:
The `Add` method also checks for `ProtoPreconditions.CheckNotNullUnconstrained(item, nameof(item));`. Are we sure that the read item cannot be null? (maybe add a comment?).

Reply:
It can't return null. I've added a comment explaining why.